### DISCUSSION TITLE
supabase-js: joining tables based on custom foreign key

### DISCUSSION THREAD

**Discussion by kheber92** [supabase/supabase#9000] [Questions] [answered]
Hi everyone,
I'm looking for a way to join two tables using a custom foreign key with supabase.js.

I have three tables:
```
tasks
- id
- root_task_id
- name
- quote_id

hourly_rates
- id
- amount

tasks_hourly_rates
- task_id
- hourly_rate_id
```

As you might guess, the tables `tasks` and `hourly_rates` are connected via `tasks_hourly_rates`.
Each task has a `root_task_id` which is basically an entry from the same table.

Now I'd like to fetch `tasks` with a specific `quote_id` including the respective `hourly_rates`, based on the task's `root_task_id`, not the task's `id`.

This is easy with SQL:
```
SELECT * FROM tasks
   LEFT JOIN tasks_hourly_rates ON tasks_hourly_rates.task_id = tasks.root_task_id
   LEFT JOIN hourly_rates ON hourly_rates.id = tasks_hourly_rates.hourly_rate_id
WHERE tasks.quote_id = 1
```

Although I didn't manage to find a way to find the right syntax with supabase-js.
Doing something like this will return data based on the `task_id` relationship, not `root_task_id` relationship:
```
supabase.from(`tasks`)
   .select('*, tasks_hourly_rates(*)')
   .match({ quote_id: 1 });
```

Is there any way to define the custom field / key for joining tables? 
In my case I would like to tell "use root_task_id instead of task_id to join my tables".

Thank you in advance!

**Comment by steve-chavez** [comment, marked as answer]
Hello,

So tasks has a:

- recursive relationship 
- many-to-many relationship to hourly_rates
- one-to-many to tasks_hourly_rates

Since the one-to-many relationship to `tasks_hourly_rates` is defined by `task.id - tasks_hourly_rates.task_id` this will be the JOIN condition used. `root_task_id` is not included because it's not part of the foreign key constraint columns.

> Is there any way to define the custom field / key for joining tables

It's possible with https://postgrest.org/en/v10.0/api.html#computed-relationships. This version is not yet deployed on supabase instances but if you make a request to support we can upgrade you to v10.

An alternative is to use a function plus [rpc](https://supabase.com/docs/reference/javascript/rpc).

**Reply by yao-dev** [reply]
Hi @steve-chavez ,

I have a similar issue detailed here https://github.com/supabase/supabase/discussions/11440 could you help me with this one? Thanks
